File: backend/services/password.py
"""Password policy for account signup.

Pure, no DB/network access — trivially testable (see tests/test_password.py).

Policy: 8-20 characters, and at least 2 of 3 character classes present
(letter / digit / symbol). "Letter" means any Unicode letter, not just ASCII,
so Korean passphrases pass; whitespace belongs to no class.

Single source of truth for the rule: schemas.py deliberately has no
pydantic-level length constraint (that would surface as a "Value error, ..."
422 leaking into the Korean UI). routers/users.py calls validate_password()
explicitly for a clean 400 instead.

frontend/src/utils/password.ts mirrors this rule by hand for the live
signup checklist.
"""
from typing import Optional

MIN_LENGTH = 8
MAX_LENGTH = 20

_LENGTH_ERROR = f"비밀번호는 {MIN_LENGTH}자 이상 {MAX_LENGTH}자 이하여야 합니다."
_VARIETY_ERROR = "비밀번호는 영문·숫자·기호 중 2종 이상을 포함해야 합니다."


def _classes(password: str) -> set:
    """Which of {"letter", "digit", "symbol"} appear in password.

    A character contributes to at most one class; whitespace contributes to
    none (see module docstring).
    """
    found = set()
    for ch in password:
        if ch.isalpha():
            found.add("letter")
        elif ch.isdigit():
            found.add("digit")
        elif not ch.isspace():
            found.add("symbol")
    return found


def validate_password(password: str) -> Optional[str]:
    """Return a Korean error message, or None if the password is acceptable."""
    if len(password) < MIN_LENGTH or len(password) > MAX_LENGTH:
        return _LENGTH_ERROR
    if len(_classes(password)) < 2:
        return _VARIETY_ERROR
    return None

File: frontend/src/utils/password.ts
/**
 * Password policy for account signup — mirrors backend/services/password.py.
 * See that module's docstring for the full reasoning (why "letter" means any
 * Unicode letter, why whitespace counts as no class, why length is capped).
 * The two are kept in sync by hand; there is no shared source of truth.
 *
 * Policy: 8-20 characters, and at least 2 of 3 character classes present
 * (letter / digit / symbol).
 */

export const MIN_LENGTH = 8;
export const MAX_LENGTH = 20;

export type PasswordChecks = {
  length: boolean;
  variety: boolean;
};

const LENGTH_ERROR = `비밀번호는 ${MIN_LENGTH}자 이상 ${MAX_LENGTH}자 이하여야 합니다.`;
const VARIETY_ERROR = '비밀번호는 영문·숫자·기호 중 2종 이상을 포함해야 합니다.';

const LETTER_RE = /\p{L}/u;
const DIGIT_RE = /\p{N}/u;
const SPACE_RE = /\s/;

function classesOf(password: string): Set<'letter' | 'digit' | 'symbol'> {
  const found = new Set<'letter' | 'digit' | 'symbol'>();
  for (const ch of password) {
    if (LETTER_RE.test(ch)) {
      found.add('letter');
    } else if (DIGIT_RE.test(ch)) {
      found.add('digit');
    } else if (!SPACE_RE.test(ch)) {
      found.add('symbol');
    }
  }
  return found;
}

// Drives the live signup checklist — each field ticks independently as the
// user types, rather than collapsing straight to a single pass/fail.
export function checkPassword(password: string): PasswordChecks {
  return {
    length: password.length >= MIN_LENGTH && password.length <= MAX_LENGTH,
    variety: classesOf(password).size >= 2,
  };
}

// Drives submit-time validation — a single Korean error message, or null if
// the password is acceptable.
export function passwordError(password: string): string | null {
  const checks = checkPassword(password);
  if (!checks.length) {
    return LENGTH_ERROR;
  }
  if (!checks.variety) {
    return VARIETY_ERROR;
  }
  return null;
}
